### PR TITLE
Respect target type in ov_add_version_defines

### PR DESCRIPTION
**Pull request by lmielick**
### Details:
Make ov_add_version_defines macro respect the original library type (instead of implicit assumption based on BUILD_SHARED_LIBS).
Note adding objects into a library with $<TARGET_OBJECTS:> does not work when mixing STATIC and OBJECT targets.

### CODE REVIEW

**Comment by ilya-lavrenov** [comment]
Object libraries are required for overcome limitation of `Makefiles` cmake generator, which does not support flags, definitions per source file and it leads to rebuild of everything. See https://github.com/openvinotoolkit/openvino/pull/16729

See also https://stackoverflow.com/questions/39525528/trying-to-trigger-a-minimal-rebuild-with-cmake-and-set-source-files-properties

**Comment by lmielick** [comment]
> Object libraries are required for overcome limitation of `Makefiles` cmake generator, which does not support flags, definitions per source file and it leads to rebuild of everything. See #16729
> 
> See also https://stackoverflow.com/questions/39525528/trying-to-trigger-a-minimal-rebuild-with-cmake-and-set-source-files-properties

Interesting, didn't test it with make, but Cmake accepts this implementation.

The primary problem I hit with the original implementation is that it decides to use STATIC or OBJECT library based BUILD_SHARED_LIBS, but I must use STATIC for the plugin target in either case. `target_sources(${TARGET} PRIVATE $<TARGET_OBJECTS:${TARGET}_version>)` does not work if `${TARGET}_version` is OBJECT and `${TARGET}` is STATIC.

**Comment by ilya-lavrenov** [comment]
> Interesting, didn't test it with make, but Cmake accepts this implementation.

Yes, it accepts. And then cmake generates Makefiles projects, which rebuild a lot lot files when product version is changed (while actually you can pull a single commit with changes in documentation)

**Comment by lmielick** [comment]
> Maybe you can extract target type using `get_target_property(target_type ${TARGET} TYPE)` and make a decision based on `target_type`?

Yes, thought of the same. Repurposed this PR for such solution.
Can't change the branch name though.

File: cmake/developer_package/version.cmake [hunk 1]
```diff
@@ -160,13 +160,17 @@ macro (ov_add_version_defines FILE TARGET)
...
+    endif()
+    set(__lib_type ${CMAKE_MATCH_1})
     _remove_source_from_target(${TARGET} ${FILE})
     _remove_source_from_target(${TARGET} ${__version_file})
-    if (BUILD_SHARED_LIBS)
-        add_library(${TARGET}_version OBJECT ${__version_file})
-    else()
-        add_library(${TARGET}_version STATIC ${__version_file})
-    endif()
+    add_library(${TARGET}_version ${__lib_type} ${__version_file})
```
Comment (ilya-lavrenov) [inline cmake/developer_package/version.cmake:173]: what if `__lib_type` is `SHARED`? I suppose we don't need shared library with version :)

Comment (lmielick) [reply cmake/developer_package/version.cmake:173]: I guess I can add exception for this, though not sure the original code was resilient to this.
It appears we use MODULE so at least MODULE/OBJECT/STATIC need to be accepted.

Comment (lmielick) [reply cmake/developer_package/version.cmake:173]: On the other hand not sure we need to block SHARED as it should work in principle.
Cmake allows library with no explicit type build to STATIC or SHARED based on BUILD_SHARED_LIBS.
Not sure we want to make it break in that flow even if we don't rely on that.

Comment (ilya-lavrenov) [reply cmake/developer_package/version.cmake:173]: version library is an internal artifact. If you create shared library with it => you expose it to end users.
We must create only static / object libs here, all other types are not accepted

Comment (lmielick) [reply cmake/developer_package/version.cmake:174]: Updated

**Comment by ilya-lavrenov** [comment]
build_jenkins

**Comment by ilya-lavrenov** [comment]
build_jenkins

**Review by ilya-lavrenov — APPROVED** [review]

**Pull request merged**